compiler/rustc_hir_typeck/src/gather_locals.rs

```diff
@@ -218,7 +218,12 @@ impl<'a, 'tcx> Visitor<'tcx> for GatherLocalsVisitor<'a, 'tcx> {
             );
         }
         let old_outermost_fn_param_pat = self.outermost_fn_param_pat.take();
-        intravisit::walk_pat(self, p);
+        if let PatKind::Guard(subpat, _) = p.kind {
+            // We'll visit the guard when checking it. Don't gather its locals twice.
```

Comment:
Sorry, I don't understand this. When is the guard expression checked, if not here?

Reply:
It's checked in `check_pat_inner`; see the `check_expr_has_type_or_error` call there.